File: temas/herencia/python/ejemplo.py
class Animal:
    def __init__(self, nombre):
        self.nombre = nombre

    def mostrar_nombre(self):
        return self.nombre


class Volador:
    def __init__(self, nombre, raza):
        self.raza = raza
        self.nombre = nombre
    
    def mostrar_nombre(self):
        return self.nombre

    def volar(self):
        return "El animal puede volar"

    def mostrar_raza(self):
        return self.raza


class Nadador:
    def __init__(self, nombre, raza):
        self.raza = raza
        self.nombre = nombre
        
    def mostrar_nombre(self):
        return self.nombre
        
    def nadar(self):
        return "El animal puede nadar"

    def mostrar_raza(self):
        return self.raza


class Ave(Animal, Volador):
    def __init__(self, nombre, raza):
        Animal.__init__(self, nombre)
        Volador.__init__(self, nombre, raza)

    def accion(self):
        return "Soy un ave" + "\n" + Volador.volar(self)


class Pez(Nadador, Animal):
    def __init__(self, nombre, raza):
        Animal.__init__(self, nombre)
        Nadador.__init__(self, nombre, raza)

File: temas/abstraccion/scala/proyecto/src/main/scala/ejemplo.scala
package ejemplo

// Definición de los traits
trait Volador {
  def fecha(): String
  def volar(): String = "¡Estoy volando!"
  //def saludo(): String = "Hola, sé volar"
}

trait Nadador {
  def nadar(): String = "¡Estoy nadando!"
  //def saludo(): String = "Hola, sé nadar"
}

trait Caminante {
  def caminar(): String = "¡Estoy caminando!"
  def saludo(): String = "Hola, sé andar"

}

// Clase que hereda de los tres traits
class Animal extends Volador with Nadador with Caminante {
  override def fecha(): String = "¡Hoy es jueves!"
}

class Mamifero {
  def reproduccion(): String = "Soy mamifero!"
}

// Objeto singleton Main
object Main {
  def main(args: Array[String]): Unit = {
    val animal = new Animal
    val ballena = new Mamifero with Nadador

    println(animal.volar())    // El animal puede volar
    println(animal.nadar())    // El animal puede nadar
    println(animal.caminar())  // El animal puede caminar
    println(animal.fecha())

    println(ballena.nadar())
    println(ballena.reproduccion())
  }
}
